src/common/src/util/value_encoding/mod.rs

```diff
@@ -89,50 +97,63 @@ fn serialize_value(value: ScalarRefImpl<'_>, mut buf: impl BufMut) {
         ScalarRefImpl::NaiveTime(v) => {
             serialize_naivetime(v.0.num_seconds_from_midnight(), v.0.nanosecond(), buf)
         }
-        ScalarRefImpl::Struct(s) => {
-            serialize_struct_or_list(s.to_protobuf_owned(), buf);
-        }
-        ScalarRefImpl::List(list) => {
-            serialize_struct_or_list(list.to_protobuf_owned(), buf);
-        }
+        ScalarRefImpl::Struct(s) => serialize_struct(s, buf),
+        ScalarRefImpl::List(list) => serialize_list(list, buf),
     }
 }
 
-fn serialize_struct_or_list(bytes: Vec<u8>, mut buf: impl BufMut) {
-    buf.put_u32_le(bytes.len() as u32);
-    buf.put_slice(bytes.as_slice());
+fn serialize_struct(value: StructRef<'_>, buf: &mut impl BufMut) {
+    value
+        .fields_ref()
+        .iter()
+        .map(|field_value| {
+            serialize_datum_ref(field_value, buf);
+        })
+        .collect_vec();
+}
+
+fn serialize_list(value: ListRef<'_>, buf: &mut impl BufMut) {
+    let values_ref = value.values_ref();
+    buf.put_u32_le(values_ref.len() as u32);
+
+    values_ref
+        .iter()
+        .map(|field_value| {
+            serialize_datum_ref(field_value, buf);
+        })
+        .collect_vec();
 }
 
-fn serialize_str(bytes: &[u8], mut buf: impl BufMut) {
+fn serialize_str(bytes: &[u8], buf: &mut impl BufMut) {
     buf.put_u32_le(bytes.len() as u32);
     buf.put_slice(bytes);
 }
 
-fn serialize_interval(interval: &IntervalUnit, mut buf: impl BufMut) {
+fn serialize_interval(interval: &IntervalUnit, buf: &mut impl BufMut) {
     buf.put_i32_le(interval.get_months());
     buf.put_i32_le(interval.get_days());
     buf.put_i64_le(interval.get_ms());
 }
 
-fn serialize_naivedate(days: i32, mut buf: impl BufMut) {
+fn serialize_naivedate(days: i32, buf: &mut impl BufMut) {
     buf.put_i32_le(days);
 }
 
-fn serialize_naivedatetime(secs: i64, nsecs: u32, mut buf: impl BufMut) {
+fn serialize_naivedatetime(secs: i64, nsecs: u32, buf: &mut impl BufMut) {
     buf.put_i64_le(secs);
     buf.put_u32_le(nsecs);
 }
 
-fn serialize_naivetime(secs: u32, nano: u32, mut buf: impl BufMut) {
+fn serialize_naivetime(secs: u32, nano: u32, buf: &mut impl BufMut) {
     buf.put_u32_le(secs);
     buf.put_u32_le(nano);
 }
 
-fn serialize_decimal(decimal: &Decimal, mut buf: impl BufMut) {
+fn serialize_decimal(decimal: &Decimal, buf: &mut impl BufMut) {
     buf.put_slice(&decimal.unordered_serialize());
 }
 
-fn deserialize_value(ty: &DataType, mut data: impl Buf) -> Result<Datum> {
+fn deserialize_value(ty: &DataType, data: &mut impl Buf) -> Result<Datum> {
```

Comment:
I guess you can pass a mutable reference of slice/cursor as `impl Buf`. `&mut impl Buf` sounds weird. 🤔

Reply:
https://docs.rs/bytes/latest/bytes/trait.Buf.html#impl-Buf-for-%26mut%20T

Reply:
I also find it strange to use `&mut impl Buf`, but using reference of slice/cursor needs to change the signature of `serialize/deserialize_datum`.

Reply:
`&mut impl Buf` is also `impl Buf`. 😄 So there might be nothing to change.